Mapper bean could not be created when using Mapping annotation with Kotlin

### Expected Behavior

A mapper bean is created and can map an entity class to a DTO

### Actual Behaviour

Exception: `io.micronaut.context.exceptions.NoSuchBeanException` is thrown. No bean of type [xxx] exists

### Steps To Reproduce

The problem only exists if we use `@Mapping`, however, if we just use `@Mapper` it works.

I've created an example in this repo: https://github.com/steventen/micronaut-bean-mapper-example

I have a `CompanyMapper` in the [Mapper](https://github.com/steventen/micronaut-bean-mapper-example/blob/main/src/main/kotlin/com/example/Mapper.kt) file, which looks like this:
```Kotlin
@Singleton
abstract class CompanyMapper {
    @Mapping(to = "companyName", from = "#{company.name}")
    abstract fun toDto(company: Company): CompanyDto
}
```
Where the entity and dto are defined as:
```Kotlin
@Entity
@Table(name = "companies")
@Introspected
data class Company(
    @Id
    val id: String,

    @Column(name = "company_name")
    val name: String,

    @Column(name = "company_value")
    val value: BigDecimal?,
)

@Serdeable
data class CompanyDto(
    val companyName: String,
)
```

To reproduce, pull the code from the repo, then
1. start the database: docker-compose up
2. start the server: ./gradlew run
3. visit the endpoint: http://localhost:8080/companies/a

When I call the api endpoint (http://localhost:8080/companies/a) which retrieves a Company record and map it to CompanyDto, it causes the following exception:
```
Message: No bean of type [com.example.CompanyMapper] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).
Path Taken: new CompanyController(CompanyRepository companyRepository,CompanyMapper companyMapper) --> new CompanyController(CompanyRepository companyRepository,[CompanyMapper companyMapper])
        at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2181)
        at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1328)
        at com.example.$CompanyController$Definition.instantiate(Unknown Source)
        at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2309)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2279)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2291)
        at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3054)
        at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
        at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2956)
        at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2917)
        at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2891)
        at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1695)
        at io.micronaut.context.DefaultBeanContext$BeanContextExecutionHandle.getTarget(DefaultBeanContext.java:4247)
        at io.micronaut.context.DefaultBeanContext$BeanContextUnsafeExecutionHandle.invokeUnsafe(DefaultBeanContext.java:4214)
        at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:228)
        at io.micronaut.http.context.ServerRequestContext.with(ServerRequestContext.java:74)
        at io.micronaut.http.server.RouteExecutor.executeRouteAndConvertBody(RouteExecutor.java:480)
        at io.micronaut.http.server.RouteExecutor.callRoute(RouteExecutor.java:470)
        at io.micronaut.http.server.RequestLifecycle.lambda$normalFlow$2(RequestLifecycle.java:146)
        at io.micronaut.core.execution.ImperativeExecutionFlowImpl.flatMap(ImperativeExecutionFlowImpl.java:72)
        at io.micronaut.http.server.RequestLifecycle.lambda$normalFlow$4(RequestLifecycle.java:146)
        at io.micronaut.http.server.RequestLifecycle.lambda$runWithFilters$14(RequestLifecycle.java:264)
        at io.micronaut.http.filter.TerminalFilter.processRequestFilter(TerminalFilter.java:58)
        at io.micronaut.http.filter.FilterRunner.filterRequest0(FilterRunner.java:153)
        at io.micronaut.http.filter.FilterRunner.filterRequest0(FilterRunner.java:153)
        at io.micronaut.http.filter.FilterRunner.lambda$filterRequest0$2(FilterRunner.java:153)
        at io.micronaut.core.execution.ImperativeExecutionFlowImpl.flatMap(ImperativeExecutionFlowImpl.java:72)
        at io.micronaut.http.filter.MethodFilter.processRequestFilter(MethodFilter.java:263)
        at io.micronaut.http.filter.FilterRunner.filterRequest0(FilterRunner.java:153)
        at io.micronaut.http.filter.FilterRunner.filterRequest(FilterRunner.java:137)
        at io.micronaut.http.filter.FilterRunner.run(FilterRunner.java:132)
        at io.micronaut.http.server.RequestLifecycle.runWithFilters(RequestLifecycle.java:281)
        at io.micronaut.http.server.RequestLifecycle.normalFlow(RequestLifecycle.java:143)
        at io.micronaut.http.server.netty.NettyRequestLifecycle.handleNormal(NettyRequestLifecycle.java:85)
        at io.micronaut.http.server.netty.RoutingInBoundHandler.accept(RoutingInBoundHandler.java:220)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler$MessageInboundHandler.read(PipeliningServerHandler.java:357)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler.channelRead(PipeliningServerHandler.java:206)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [com.example.CompanyMapper] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).
        at io.micronaut.context.DefaultBeanContext.newNoSuchBeanException(DefaultBeanContext.java:2773)
        at io.micronaut.context.DefaultApplicationContext.newNoSuchBeanException(DefaultApplicationContext.java:304)
        at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2735)
        at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1729)
        at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:89)
        at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2165)
```

Also tried to use `interface` and the problem is the same.

### Environment Information

- MacOS 14.2.1
- JDK 17

### Example Application

https://github.com/steventen/micronaut-bean-mapper-example

### Version

4.2.3

--- a/core-processor/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java
+++ b/core-processor/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java
@@ -1052,10 +1052,12 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
         if (AnnotationUtil.INTERNAL_ANNOTATION_NAMES.contains(annotationValue.getAnnotationName()) || context.isProcessed(annotationValue)) {
             return Stream.empty();
         }
-        String repeatableContainer = getRepeatableContainerNameForType(processedAnnotation.getAnnotationType());
-        if (repeatableContainer != null) {
-            // Collect the repeatable container from the annotation mirror for later
-            context.repeatableToContainer.put(annotationValue.getAnnotationName(), repeatableContainer);
+        if (processedAnnotation.getAnnotationType() != null) {
+            String repeatableContainer = getRepeatableContainerNameForType(processedAnnotation.getAnnotationType());
+            if (repeatableContainer != null) {
+                // Collect the repeatable container from the annotation mirror for later
+                context.repeatableToContainer.put(annotationValue.getAnnotationName(), repeatableContainer);
+            }
         }
 
         // Add annotation default values

--- a/core-processor/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java
+++ b/core-processor/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java
@@ -523,7 +523,7 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
      * @return Return the name or null
      */
     @Nullable
-    protected abstract String getRepeatableNameForType(T annotationType);
+    protected abstract String getRepeatableContainerNameForType(T annotationType);
 
     /**
      * @param annotationElement The annotation element
@@ -984,14 +984,15 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
                 Map.entry(
                         List.of(), annotationValues
                 )
-        ));
+        ), processingContext.repeatableToContainer);
     }
 
     private void addAnnotations(@NonNull MutableAnnotationMetadata annotationMetadata,
                                 boolean isDeclared,
                                 boolean isStereotype,
                                 boolean alwaysIncludeAnnotation,
-                                @NonNull List<Map.Entry<List<String>, List<AnnotationValue<?>>>> annotations) {
+                                @NonNull List<Map.Entry<List<String>, List<AnnotationValue<?>>>> annotations,
+                                @NonNull Map<String, String> repeatableToContainer) {
 
         // We need to add annotations by their levels:
         // 1. The annotation
@@ -1014,7 +1015,8 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
                             parentAnnotations,
                             isDeclared,
                             isStereotype,
-                            annotationValue);
+                            annotationValue,
+                            repeatableToContainer);
 
                     List<String> newParentAnnotations = CollectionUtils.concat(parentAnnotations, annotationValue.getAnnotationName());
 
@@ -1031,7 +1033,7 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
         }
 
         if (!stereotypes.isEmpty()) {
-            addAnnotations(annotationMetadata, isDeclared, true, alwaysIncludeAnnotation, stereotypes);
+            addAnnotations(annotationMetadata, isDeclared, true, alwaysIncludeAnnotation, stereotypes, repeatableToContainer);
         }
     }
 
@@ -1050,6 +1052,12 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
         if (AnnotationUtil.INTERNAL_ANNOTATION_NAMES.contains(annotationValue.getAnnotationName()) || context.isProcessed(annotationValue)) {
             return Stream.empty();
         }
+        String repeatableContainer = getRepeatableContainerNameForType(processedAnnotation.getAnnotationType());
+        if (repeatableContainer != null) {
+            // Collect the repeatable container from the annotation mirror for later
+            context.repeatableToContainer.put(annotationValue.getAnnotationName(), repeatableContainer);
+        }
+
         // Add annotation default values
         processedAnnotation = addDefaults(processedAnnotation);
         // Check if the annotation has the stereotypes set manually, before adding alias stereotypes
@@ -1205,7 +1213,7 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
         boolean isRepeatableAnnotationContainer = !repeatableAnnotations.isEmpty() && repeatableAnnotations.stream()
                 .allMatch(value -> {
                     T annotationMirror = getAnnotationMirror(value.getAnnotationName()).orElse(null);
-                    return annotationMirror != null && getRepeatableNameForType(annotationMirror) != null;
+                    return annotationMirror != null && getRepeatableContainerNameForType(annotationMirror) != null;
                 });
         if (isRepeatableAnnotationContainer) {
             // Repeatable annotations container is being added with values
@@ -1262,7 +1270,8 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
                                @NonNull List<String> parentAnnotations,
                                boolean isDeclared,
                                boolean isStereotype,
-                               @NonNull AnnotationValue<?> annotationValue) {
+                               @NonNull AnnotationValue<?> annotationValue,
+                               @NonNull Map<String, String> repeatableToContainer) {
 
         String annotationName = annotationValue.getAnnotationName();
         Map<CharSequence, Object> annotationDefaults = annotationValue.getDefaultValues();
@@ -1272,8 +1281,14 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
             throw new IllegalStateException("Annotation should contain default values and an empty list " + annotationValue.getAnnotationName());
         }
 
-        T annotationMirror = getAnnotationMirror(annotationName).orElse(null);
-        String repeatableContainer = annotationMirror != null ? getRepeatableNameForType(annotationMirror) : null;
+        String repeatableContainer = repeatableToContainer.get(annotationName);
+        if (repeatableContainer == null) {
+            repeatableContainer = AnnotationMetadataSupport.getCoreRepeatableAnnotationsContainers().get(annotationName);
+        }
+        if (repeatableContainer == null) {
+            T annotationMirror = getAnnotationMirror(annotationName).orElse(null);
+            repeatableContainer = annotationMirror != null ? getRepeatableContainerNameForType(annotationMirror) : null;
+        }
         if (isStereotype) {
             if (repeatableContainer != null) {
                 if (isDeclared) {
@@ -1596,7 +1611,7 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
         return modify(annotationMetadata, metadata -> {
             T annotationMirror = getAnnotationMirror(annotationType).orElse(null);
             if (annotationMirror != null) {
-                String repeatableName = getRepeatableNameForType(annotationMirror);
+                String repeatableName = getRepeatableContainerNameForType(annotationMirror);
                 if (repeatableName != null) {
                     metadata.removeAnnotation(repeatableName);
                 } else {
@@ -1622,7 +1637,7 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
         return modify(annotationMetadata, metadata -> {
             T annotationMirror = getAnnotationMirror(annotationType).orElse(null);
             if (annotationMirror != null) {
-                String repeatableName = getRepeatableNameForType(annotationMirror);
+                String repeatableName = getRepeatableContainerNameForType(annotationMirror);
                 if (repeatableName != null) {
                     metadata.removeStereotype(repeatableName);
                 } else {
@@ -1677,14 +1692,16 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
      * @param visitorContext    The visitor context
      * @param parentAnnotations The parent annotations
      * @param processedVisitors The processed visitors
+     * @param repeatableToContainer The repeatable to container
      * @since 4.0.0
      */
     private record ProcessingContext(@NonNull VisitorContext visitorContext,
                                      @NonNull Set<String> parentAnnotations,
-                                     @NonNull Set<Class<?>> processedVisitors) {
+                                     @NonNull Set<Class<?>> processedVisitors,
+                                     @NonNull Map<String, String> repeatableToContainer) {
 
         ProcessingContext(@NonNull VisitorContext visitorContext) {
-            this(visitorContext, Collections.emptySet(), Collections.emptySet());
+            this(visitorContext, Collections.emptySet(), Collections.emptySet(), new HashMap<>());
         }
 
         boolean isProcessed(@NonNull AnnotationValue<?> annotationValue) {
@@ -1694,19 +1711,19 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
         @NonNull
         ProcessingContext withParent(@NonNull AnnotationValue<?> parent) {
             Set<String> parents = CollectionUtils.concat(parentAnnotations, parent.getAnnotationName());
-            return new ProcessingContext(visitorContext, Collections.unmodifiableSet(parents), processedVisitors);
+            return new ProcessingContext(visitorContext, Collections.unmodifiableSet(parents), processedVisitors, repeatableToContainer);
         }
 
         @NonNull
         ProcessingContext withParents(@NonNull List<String> newParents) {
             Set<String> parents = CollectionUtils.concat(parentAnnotations, newParents);
-            return new ProcessingContext(visitorContext, Collections.unmodifiableSet(parents), processedVisitors);
+            return new ProcessingContext(visitorContext, Collections.unmodifiableSet(parents), processedVisitors, repeatableToContainer);
         }
 
         @NonNull
         public ProcessingContext withProcessedVisitor(@NonNull Class<?> processedVisitor) {
             Set<Class<?>> visitors = CollectionUtils.concat(processedVisitors, processedVisitor);
-            return new ProcessingContext(visitorContext, parentAnnotations, Collections.unmodifiableSet(visitors));
+            return new ProcessingContext(visitorContext, parentAnnotations, Collections.unmodifiableSet(visitors), repeatableToContainer);
         }
     }
 

--- a/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/annotation/GroovyAnnotationMetadataBuilder.java
+++ b/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/annotation/GroovyAnnotationMetadataBuilder.java
@@ -244,11 +244,11 @@ public class GroovyAnnotationMetadataBuilder extends AbstractAnnotationMetadataB
 
     @Override
     protected String getRepeatableName(AnnotationNode annotationMirror) {
-        return getRepeatableNameForType(annotationMirror.getClassNode());
+        return getRepeatableContainerNameForType(annotationMirror.getClassNode());
     }
 
     @Override
-    protected String getRepeatableNameForType(AnnotatedNode annotationType) {
+    protected String getRepeatableContainerNameForType(AnnotatedNode annotationType) {
         List<AnnotationNode> annotationNodes = annotationType.getAnnotations(ClassHelper.makeCached(Repeatable.class));
         if (CollectionUtils.isNotEmpty(annotationNodes)) {
             Expression expression = annotationNodes.get(0).getMember("value");
@@ -318,7 +318,7 @@ public class GroovyAnnotationMetadataBuilder extends AbstractAnnotationMetadataB
             if (value instanceof ListExpression listExpression) {
                 for (Expression expression : listExpression.getExpressions()) {
                     if (expression instanceof AnnotationConstantExpression annotationConstantExpression) {
-                        String name = getRepeatableNameForType(expression.getType());
+                        String name = getRepeatableContainerNameForType(expression.getType());
                         if (name != null && name.equals(node.getClassNode().getName())) {
                             repeatable = true;
                             expanded.add((AnnotationNode) annotationConstantExpression.getValue());

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/JavaAnnotationMetadataBuilder.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/JavaAnnotationMetadataBuilder.java
@@ -132,12 +132,12 @@ public class JavaAnnotationMetadataBuilder extends AbstractAnnotationMetadataBui
     @Override
     protected String getRepeatableName(AnnotationMirror annotationMirror) {
         final Element typeElement = annotationMirror.getAnnotationType().asElement();
-        return getRepeatableNameForType(typeElement);
+        return getRepeatableContainerNameForType(typeElement);
     }
 
     @Nullable
     @Override
-    protected String getRepeatableNameForType(Element annotationType) {
+    protected String getRepeatableContainerNameForType(Element annotationType) {
         List<? extends AnnotationMirror> mirrors = annotationType.getAnnotationMirrors();
         for (AnnotationMirror mirror : mirrors) {
             String name = mirror.getAnnotationType().toString();

--- a/inject-java/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
@@ -2,6 +2,7 @@ package io.micronaut.inject.beans
 
 import io.micronaut.annotation.processing.test.AbstractTypeElementSpec
 import io.micronaut.context.ApplicationContext
+import io.micronaut.context.annotation.Mapper
 import io.micronaut.core.annotation.AnnotationUtil
 import io.micronaut.core.annotation.Order
 import io.micronaut.core.reflect.ClassUtils
@@ -788,5 +789,49 @@ interface MyEntityControllerInterface {
 
         expect:
             definition == null
+    }
+
+    void "test mapped interface bean"() {
+        given:
+            def definition = buildSimpleInterceptedBeanDefinition('test.ProductMappers', '''
+package test;
+
+import io.micronaut.context.annotation.Mapper.Mapping;
+import jakarta.inject.Singleton;
+import io.micronaut.core.annotation.Introspected;
+
+@Singleton
+interface ProductMappers {
+    @Mapping(
+        to = "price",
+        from = "#{product.price * 2}",
+        format = "$#.00"
+    )
+    @Mapping(
+        to = "distributor",
+        from = "#{this.getDistributor()}"
+    )
+    ProductDTO toProductDTO(Product product);
+
+    default String getDistributor() {
+        return "Great Product Company";
+    }
+}
+
+@Introspected
+record Product(
+    String name,
+    double price,
+    String manufacturer) {
+}
+
+@Introspected
+record ProductDTO(String name, String price, String distributor) {
+}
+''')
+
+        expect:
+            definition.getExecutableMethods()[0].hasDeclaredAnnotation(Mapper)
+            definition.getExecutableMethods()[0].hasDeclaredAnnotation(Mapper.Mapping)
     }
 }

--- a/inject-kotlin-test/src/main/groovy/io/micronaut/annotation/processing/test/KotlinCompiler.java
+++ b/inject-kotlin-test/src/main/groovy/io/micronaut/annotation/processing/test/KotlinCompiler.java
@@ -177,6 +177,10 @@ public class KotlinCompiler {
         return loadReference(name, clazz, BeanDefinitionWriter.CLASS_SUFFIX + BeanDefinitionReferenceWriter.REF_SUFFIX);
     }
 
+    public static BeanDefinition<?> buildIntroducedBeanDefinition(String className, @Language("kotlin") String cls) throws InstantiationException, NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        return loadReference(className, cls, BeanDefinitionVisitor.PROXY_SUFFIX + BeanDefinitionWriter.CLASS_SUFFIX);
+    }
+
     public static BeanDefinition<?> buildInterceptedBeanDefinition(String className, @Language("kotlin") String cls) throws InstantiationException, NoSuchMethodException, InvocationTargetException, IllegalAccessException {
         return loadReference(className, cls, BeanDefinitionWriter.CLASS_SUFFIX + BeanDefinitionVisitor.PROXY_SUFFIX + BeanDefinitionWriter.CLASS_SUFFIX);
     }

--- a/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/annotation/KotlinAnnotationMetadataBuilder.kt
+++ b/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/annotation/KotlinAnnotationMetadataBuilder.kt
@@ -479,12 +479,12 @@ internal class KotlinAnnotationMetadataBuilder(
     }
 
     override fun getRepeatableName(annotationMirror: KSAnnotation): String? {
-        return getRepeatableNameForType(annotationMirror.annotationType)
+        return getRepeatableContainerNameForType(annotationMirror.annotationType)
     }
 
-    override fun getRepeatableNameForType(annotationType: KSAnnotated): String? {
+    override fun getRepeatableContainerNameForType(annotationType: KSAnnotated): String? {
         val name = java.lang.annotation.Repeatable::class.java.name
-        val repeatable = annotationType.getClassDeclaration(visitorContext).annotations.find {
+        val repeatable = annotationType.annotations.find {
             it.annotationType.resolve().declaration.qualifiedName?.asString() == name
         }
         if (repeatable != null) {

--- a/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/annotation/KotlinAnnotationMetadataBuilder.kt
+++ b/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/annotation/KotlinAnnotationMetadataBuilder.kt
@@ -479,7 +479,7 @@ internal class KotlinAnnotationMetadataBuilder(
     }
 
     override fun getRepeatableName(annotationMirror: KSAnnotation): String? {
-        return getRepeatableContainerNameForType(annotationMirror.annotationType)
+        return getRepeatableContainerNameForType(annotationMirror.annotationType.getClassDeclaration(visitorContext))
     }
 
     override fun getRepeatableContainerNameForType(annotationType: KSAnnotated): String? {

--- a/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/KotlinClassElement.kt
+++ b/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/KotlinClassElement.kt
@@ -636,6 +636,9 @@ internal open class KotlinClassElement(
 
         @OptIn(KspExperimental::class)
         override fun getElementName(element: KSNode): String {
+            if (element is KSPropertyDeclaration) {
+                return element.simpleName.asString()
+            }
             if (element is KSFunctionDeclaration) {
                 return visitorContext.resolver.getJvmName(element)!!
             }

--- a/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/beans/BeanDefinitionSpec.groovy
+++ b/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/beans/BeanDefinitionSpec.groovy
@@ -1,6 +1,7 @@
 package io.micronaut.kotlin.processing.beans
 
 import io.micronaut.annotation.processing.test.KotlinCompiler
+import io.micronaut.context.annotation.Mapper
 import io.micronaut.context.annotation.Requires
 import io.micronaut.context.exceptions.NoSuchBeanException
 import io.micronaut.core.annotation.*
@@ -128,6 +129,35 @@ class F
         expect:
         definition.injectedMethods.size() == 0
         definition.injectedFields.size() == 1
+    }
+
+    void "test abstract mapper singleton"() {
+        given:
+        def definition = KotlinCompiler.buildIntroducedBeanDefinition('test.ProductMappers', '''
+package test
+
+import io.micronaut.core.annotation.Introspected
+import io.micronaut.context.annotation.Mapper.Mapping
+import jakarta.inject.Singleton
+
+@Singleton
+abstract class ProductMappers {
+    @Mapping(to = "price", from = "#{product.price * 2}", format = "$#.00")
+    @Mapping(to = "distributor", from = "#{this.getDistributor()}")
+    abstract fun toProductDTO(product: Product): ProductDTO
+    fun getDistributor() : String = "Great Product Company"
+}
+
+@Introspected
+data class Product(val name: String, val price: Double, val manufacturer: String)
+
+@Introspected
+data class ProductDTO(val name: String, val price: String, val distributor: String)
+''')
+
+        expect:
+        definition.getExecutableMethods()[0].hasDeclaredAnnotation(Mapper)
+        definition.getExecutableMethods()[0].hasDeclaredAnnotation(Mapper.Mapping)
     }
 
     void "test interface bean"() {

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ContactEntity.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ContactEntity.kt
@@ -1,0 +1,8 @@
+package io.micronaut.docs.ioc.mappers
+
+import io.micronaut.core.annotation.Introspected
+import io.micronaut.core.annotation.Nullable
+//tag::class[]
+@Introspected
+data class ContactEntity(@Nullable var id: Long? = null, val firstName: String, val lastName: String)
+//end::class[]

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ContactForm.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ContactForm.kt
@@ -1,0 +1,8 @@
+package io.micronaut.docs.ioc.mappers
+
+import io.micronaut.core.annotation.Introspected
+
+//tag::class[]
+@Introspected
+data class ContactForm(val firstName: String, val lastName: String)
+//end::class[]

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ContactMappers.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ContactMappers.kt
@@ -1,0 +1,14 @@
+package io.micronaut.docs.ioc.mappers
+
+//tag::imports[]
+import io.micronaut.context.annotation.Mapper
+//end::imports[]
+import io.micronaut.context.annotation.Requires
+
+@Requires(property = "spec.name", value = "SimpleMapperSpec")
+//tag::class[]
+interface ContactMappers {
+    @Mapper
+    fun toEntity(contactForm: ContactForm) : ContactEntity
+}
+//end::class[]

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/MappersSpec.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/MappersSpec.kt
@@ -1,0 +1,26 @@
+package io.micronaut.docs.ioc.mappers
+
+import io.micronaut.context.ApplicationContext
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+
+internal class MappersSpec {
+    @Test
+    fun testMappers() {
+        ApplicationContext.run().use { context ->
+            // tag::mappers[]
+            val productMappers = context.getBean(ProductMappers::class.java)
+            val (name, price, distributor) = productMappers.toProductDTO(
+                Product(
+                    "MacBook",
+                    910.50,
+                    "Apple"
+                )
+            )
+            Assertions.assertEquals("MacBook", name)
+            Assertions.assertEquals("$1821.00", price)
+            Assertions.assertEquals("Great Product Company", distributor)
+            // end::mappers[]
+        }
+    }
+}

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/Product.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/Product.kt
@@ -1,0 +1,8 @@
+package io.micronaut.docs.ioc.mappers
+
+import io.micronaut.core.annotation.Introspected
+
+// tag::class[]
+@Introspected
+data class Product(val name: String, val price: Double, val manufacturer: String)
+// end::class[]

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ProductDTO.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ProductDTO.kt
@@ -1,0 +1,8 @@
+package io.micronaut.docs.ioc.mappers
+
+import io.micronaut.core.annotation.Introspected
+
+// tag::class[]
+@Introspected
+data class ProductDTO(val name: String, val price: String, val distributor: String)
+// end::class[]

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ProductMappers.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/ProductMappers.kt
@@ -1,0 +1,16 @@
+package io.micronaut.docs.ioc.mappers
+
+
+// tag::class[]
+import io.micronaut.context.annotation.Mapper.Mapping
+import jakarta.inject.Singleton
+
+@Singleton
+abstract class ProductMappers {
+    @Mapping(to = "price", from = "#{product.price * 2}", format = "$#.00")
+    @Mapping(to = "distributor", from = "#{this.getDistributor()}")
+    abstract fun toProductDTO(product: Product): ProductDTO
+    fun getDistributor() : String = "Great Product Company"
+}
+
+// tag::class[]

--- a/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/SimpleMapperSpec.kt
+++ b/test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/ioc/mappers/SimpleMapperSpec.kt
@@ -1,0 +1,20 @@
+package io.micronaut.docs.ioc.mappers
+
+import io.micronaut.context.ApplicationContext
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+import java.util.*
+
+class SimpleMapperSpec {
+    @Test
+    fun testSimpleMappers() {
+        ApplicationContext.run(Collections.singletonMap<String, Any>("spec.name", "SimpleMapperSpec")).use { context ->
+            // tag::mappers[]
+            val contactMappers = context.getBean(ContactMappers::class.java)
+            val entity : ContactEntity = contactMappers.toEntity(ContactForm("John", "Snow"))
+            Assertions.assertEquals("John", entity.firstName)
+            Assertions.assertEquals("Snow", entity.lastName)
+            // end::mappers[]
+        }
+    }
+}

